File: src/main/java/com/nanaios/polygonal_tech/capability/impl/ItemSlot.java
package com.nanaios.polygonal_tech.capability.impl;

import com.nanaios.polygonal_tech.capability.interfaces.IItemSlot;
import net.minecraft.world.item.ItemStack;
import org.jetbrains.annotations.NotNull;

import java.util.function.Predicate;

public class ItemSlot implements IItemSlot {
    protected final Predicate<ItemStack> validator;
    protected ItemStack stack = ItemStack.EMPTY;

    public ItemSlot(Predicate<ItemStack> validator) {
        this.validator = validator;
    }

    @Override
    public @NotNull ItemStack getStack() {
        return stack;
    }

    @Override
    public @NotNull ItemStack insertItem(@NotNull ItemStack stack, boolean simulate) {
        if (!isItemValid(stack)) return stack;
        if (this.stack.isEmpty()) {
            if (!simulate) {
                this.stack = stack.copy();
            }
            return ItemStack.EMPTY;
        }

        if (ItemStack.isSameItemSameTags(this.stack, stack)) {
            int space = getSlotLimit() - this.stack.getCount();
            int toInsert = Math.min(space, stack.getCount());
            if (!simulate) {
                this.stack.grow(toInsert);
            }
            ItemStack remainder = stack.copy();
            remainder.shrink(toInsert);
            return remainder;
        }
        return stack;
    }

    @Override
    public @NotNull ItemStack extractItem(int amount, boolean simulate) {
        if (stack.isEmpty()) return ItemStack.EMPTY;
        int toExtract = Math.min(amount, stack.getCount());
        ItemStack extracted = stack.copy();
        extracted.setCount(toExtract);
        if (!simulate) {
            stack.shrink(toExtract);
        }
        return extracted;
    }

    @Override
    public int getSlotLimit() {
        return stack.getMaxStackSize();
    }

    @Override
    public boolean isItemValid(@NotNull ItemStack stack) {
        return validator.test(stack);
    }

    public static class InputOnly extends ItemSlot {
        protected final IItemSlot base;
        public InputOnly(ItemSlot slot) {
            super(slot.validator);
            this.base = slot;
        }

        @Override
        public boolean isItemValid(@NotNull ItemStack stack) {
            return base.isItemValid(stack);
        }

        @Override
        public int getSlotLimit() {
            return base.getSlotLimit();
        }

        @Override
        public @NotNull ItemStack getStack() {
            return base.getStack();
        }

        @Override
        public @NotNull ItemStack insertItem(@NotNull ItemStack stack, boolean simulate) {
            return base.insertItem(stack, simulate);
        }

        @Override
        public @NotNull ItemStack extractItem(int amount, boolean simulate) {
            return ItemStack.EMPTY;
        }
    }

    public static class OutputOnly extends ItemSlot {
        protected final IItemSlot base;
        public OutputOnly(ItemSlot slot) {
            super(slot.validator);
            this.base = slot;
        }

        @Override
        public boolean isItemValid(@NotNull ItemStack stack) {
            return false;
        }

        @Override
        public int getSlotLimit() {
            return base.getSlotLimit();
        }

        @Override
        public @NotNull ItemStack getStack() {
            return base.getStack();
        }

        @Override
        public @NotNull ItemStack insertItem(@NotNull ItemStack stack, boolean simulate) {
            return stack;
        }

        @Override
        public @NotNull ItemStack extractItem(int amount, boolean simulate) {
            return base.extractItem(amount, simulate);
        }
    }
}

File: src/main/kotlin/com/nanaios/polygonal_tech/core/capability/item/ItemSlot.kt
package com.nanaios.polygonal_tech.core.capability.item

import com.nanaios.polygonal_tech.core.network.sync.storage.EmptySyncValueStorage
import com.nanaios.polygonal_tech.core.network.sync.storage.ISyncValueStorage
import com.nanaios.polygonal_tech.core.network.sync.type.ISyncType
import com.nanaios.polygonal_tech.core.network.sync.type.SyncType
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.world.item.ItemStack
import kotlin.math.min

open class ItemSlot(
    override val x: Int,
    override val y: Int,
    override var stack: ItemStack,
    protected val valid: (ItemStack) -> Boolean
) : IItemSlot {
    override var storage: ISyncValueStorage = EmptySyncValueStorage
    override var type: ISyncType = SyncType.NONE
    override var isDirty = false
        protected set
    override val slotLimit
        get() = 64

    protected fun getStackLimit(stack: ItemStack): Int {
        return min(slotLimit, stack.maxStackSize)
    }

    protected fun markDirty() {
        isDirty = true
        storage.onSyncValueChanged(this)
    }

    override fun insertItem(
        stack: ItemStack,
        simulate: Boolean
    ): ItemStack {
        if (stack.isEmpty) return ItemStack.EMPTY
        if (!valid(stack)) return stack

        val current = this.stack
        var limit = getStackLimit(stack)

        if (!current.isEmpty) {
            if (!ItemStack.isSameItemSameTags(stack, current)) return stack
            limit -= current.count
        }

        if (limit <= 0) return stack

        val reachedLimit = stack.count > limit

        if (!simulate) {
            if (current.isEmpty) {
                this.stack = if (reachedLimit) stack.copyWithCount(limit) else stack
            } else {
                current.grow(if (reachedLimit) limit else stack.count)
            }
            markDirty()
        }

        return if (reachedLimit) stack.copyWithCount(stack.count - limit) else ItemStack.EMPTY
    }

    override fun extractItem(amount: Int, simulate: Boolean): ItemStack {
        if (amount <= 0) return ItemStack.EMPTY

        val current = this.stack
        if (current.isEmpty) return ItemStack.EMPTY

        val toExtract = min(amount, current.maxStackSize)

        if (current.count <= toExtract) {
            if (!simulate) {
                this.stack = ItemStack.EMPTY
                markDirty()
                return current
            } else {
                return current.copy()
            }
        } else {
            if (!simulate) {
                this.stack = current.copyWithCount(current.count - toExtract)
                markDirty()
            }
            return current.copyWithCount(toExtract)
        }
    }

    override fun isItemValid(stack: ItemStack) = valid(stack)

    override fun writeBuffer(buffer: FriendlyByteBuf) {
        buffer.writeItemStack(stack,true)
    }

    override fun readBuffer(buffer: FriendlyByteBuf) {
        stack = buffer.readItem()
    }

    override fun onSync() {
        isDirty = false
    }

    override fun serializeNBT(): CompoundTag {
        val tag = CompoundTag()
        stack.save(tag)
        return tag
    }

    override fun deserializeNBT(nbt: CompoundTag) {
        stack = ItemStack.of(nbt)
    }
}